compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceDelegationLowering.kt

```diff
@@ -0,0 +1,190 @@
+
+    // Constants, object accesses, reads of immutable variables, and reads of immutable properties in the same file
+    // don't need to be cached in a field; reads of mutable properties have to be as we should ignore further assignments
+    // to those, and immutable properties in other files might become mutable without breaking ABI.
+    private fun IrExpression.canInline(currentFile: IrFile): Boolean = when (this) {
+        is IrGetValue -> !symbol.owner.let { it is IrVariable && it.isVar }
+        is IrGetField -> symbol.owner.let { it.isFinal && it.fileParent == currentFile } && receiver?.canInline(currentFile) != false
+        is IrCall -> symbol.owner.let { it.isFinalDefaultValGetter && it.fileParent == currentFile } &&
+                dispatchReceiver?.canInline(currentFile) != false && extensionReceiver?.canInline(currentFile) != false
+        else -> isTrivial()
+    }
+
+    private val IrSimpleFunction.isFinalDefaultValGetter: Boolean
+        get() = origin == IrDeclarationOrigin.DEFAULT_PROPERTY_ACCESSOR &&
+                correspondingPropertySymbol?.let { it.owner.getter == this && it.owner.setter == null } == true &&
+                modality == Modality.FINAL
+
+    private fun IrExpression.inline(oldReceiver: IrValueParameter?, newReceiver: IrValueParameter?): IrExpression = when (this) {
+        is IrGetField ->
+            IrGetFieldImpl(startOffset, endOffset, symbol, type, receiver?.inline(oldReceiver, newReceiver), origin, superQualifierSymbol)
+        is IrGetValue ->
+            IrGetValueImpl(startOffset, endOffset, type, newReceiver?.symbol.takeIf { symbol == oldReceiver?.symbol } ?: symbol, origin)
+        is IrCall ->
+            IrCallImpl(startOffset, endOffset, type, symbol, typeArgumentsCount, valueArgumentsCount, origin, superQualifierSymbol).apply {
+                dispatchReceiver = this@inline.dispatchReceiver?.inline(oldReceiver, newReceiver)
+                extensionReceiver = this@inline.extensionReceiver?.inline(oldReceiver, newReceiver)
+            }
+        else -> shallowCopy()
+    }
+
+    override fun visitClass(declaration: IrClass): IrStatement {
+        declaration.transformChildren(this, null)
+        declaration.transformDeclarationsFlat {
+            (it as? IrProperty)?.transform()
+        }
+        return declaration
+    }
+
+    private fun IrProperty.transform(): List<IrDeclaration>? {
+        if (!isDelegated || isFakeOverride) return null
+
+        val oldField = backingField
+        val delegate = oldField?.initializer?.expression
+        if (delegate !is IrPropertyReference ||
+            getter?.returnsResultOfStdlibCall == false ||
+            setter?.returnsResultOfStdlibCall == false
+        ) return null
+
+        val receiver = (delegate.dispatchReceiver ?: delegate.extensionReceiver)
+            ?.transform(this@PropertyReferenceDelegationTransformer, null)
+        backingField = receiver?.takeIf { !it.canInline(fileParent) }?.let {
```

Comment:
Even if receiver can be inlined, we still need to evaluate (and drop) it in the constructor, otherwise behavior will change because of initialization order (please add as a test):
```
var result = "Fail"

object O {
    val z = 42
    init {
        result = "OK"
    }
}

class A {
    val x by O::z
}

fun box(): String {
    A()
    return result
}
```

Reply:
done